### PR TITLE
New year, new tests

### PR DESCRIPTION
**Pull request by hrhino**
I nearly resolved to clear the scala/bug backlog this year.

Fixes scala/bug#3899
Fixes scala/bug#7686

### CODE REVIEW

File: test/files/jvm/t9668.scala [hunk 1]
```diff
@@ -0,0 +1,6 @@
+object Test extends App {
+  type Foo = String
+  object Foo
+
+  println(Foo.getClass.getInterfaces.toList)
```
Comment (som-snytt) [inline test/files/jvm/t9668.scala]: `assert(blah.toList.isEmpty)` and omit the check file.

File: test/files/run/t3899/Derived_2.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
+trait T extends Base_1 {
+	def t1(as: String*): Array[String] = {
+		varargs1(as: _*)
+	}
+	def t2(as: String*): Array[String] = {
+    // This is the bug reported in the ticket.
```
Comment (som-snytt) [inline test/files/run/t3899/Derived_2.scala:6]: 2.11 says `java.lang.ClassCastException: scala.collection.mutable.WrappedArray$ofRef cannot be cast to [Ljava.lang.String;`

Comment (som-snytt) [reply test/files/run/t3899/Derived_2.scala:6]: Possible to omit `_1` and `_2`, which imply special compilation order.

File: test/files/run/t3899/Base_1.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+public class Base_1 {
+	public String[] varargs1(String... as) {
+	  return as;
```
Comment (som-snytt) [inline test/files/run/t3899/Base_1.java]: You seem to be using tabs and/or mixed in these files. Maybe you installed a new editor you received as a holiday gift?

File: test/files/jvm/t9668.scala (hunk 1 above)
Comment (som-snytt) [reply test/files/jvm/t9668.scala]: Or is this test redundant with the neg test?

File: test/files/neg/t9668.scala [hunk 4]
```diff
@@ -0,0 +1,6 @@
+object Test {
+  type Foo = String
+  object Foo
+
+  Foo : Serializable
```
Comment (som-snytt) [inline test/files/neg/t9668.scala]: Type aliases don't have companion objects.

I still see
```
class C extends Serializable ; object C ; C : Serializable
```
Is that wrong? Maybe it is wrong. Or maybe it's a feature.

File: test/files/run/t3899/Base_1.java (hunk 3 above)
Comment (som-snytt) [reply test/files/run/t3899/Base_1.java]: Or just old retronym intellij files.

File: test/files/run/t3899/Derived_2.scala (hunk 2 above)
Comment (som-snytt) [reply test/files/run/t3899/Derived_2.scala:6]: Fixed in 2.12.3.

File: test/files/jvm/t9668.scala (hunk 1 above)
Comment (He-Pin) [reply test/files/jvm/t9668.scala]: not classOf[Foo]?

**Review by som-snytt — APPROVED** [review]
Tests are documentation!

That's coming from the testiest of them all.

File: test/files/jvm/t9668.scala (hunk 1 above)
Comment (som-snytt) [reply test/files/jvm/t9668.scala]: @hepin1989 this was testing the companion object, not the companion class [or type].

**Comment by som-snytt** [comment]
The other test is included in the fix at https://github.com/scala/scala/pull/5602

**Pull request merged**